backend/wmg/pipeline/utils.py

```diff
@@ -0,0 +1,86 @@
+def remap_anndata_normalized_X_to_raw_X_if_exists(adata: AnnData) -> None:
+    """
+    Remaps AnnData.X to AnnData.raw.X if and only if AnnData.raw.X exists.
+
+    Furthermore, garbage collect the original value pointed to by AnnData.X.
+
+    Parameters
+    ----------
+    data
+        The (annotated) data matrix of shape `n_obs` × `n_vars`.
+        Rows correspond to cells and columns to genes.
+    """
+    raw_expression_matrix = getattr(adata.raw, "X", None)
+
+    if raw_expression_matrix is not None:
+        logger.info("Remapping AnnData.X to AnnData.raw.X")
+        adata.X = adata.raw.X
+        gc.collect()
+
+
+def anndata_filter_cells_by_gene_counts_inplace(adata: AnnData, min_genes: int) -> None:
+    """
+    Filter cell outliers based on numbers of genes expressed.
+
+    NOTE: This operation is *in place*
+
+    This implementation is more memory efficient than `scanpy.pp.filter_cells`:
+    https://github.com/scverse/scanpy/blob/89804c26bcbdde4449dfe98b3193a9d28c2a0e2f/scanpy/preprocessing/_simple.py#L41C2-L41C2
+
+    The scanpy implementation is more memory intensive because it makes unnecessary copies
+    of the underlying data structures when performing computations.
+
+    Parameters
+    ----------
+    data
+        The (annotated) data matrix of shape `n_obs` × `n_vars`.
+        Rows correspond to cells and columns to genes.
+    min_genes
+        Minimum number of genes expressed required for a cell to pass filtering.
+
+    """
+
+    # memory efficient implementation for getting count of non-zeros per row.
+    # For CSR we use the `indptr` structure to compute number of non-zeros per row
+    # For CSC we use the `indices` structure to compute number of non-zeros per row
+    # And finally for ndarray, we simply filter the matrix and sum the resulting boolean matrix.
+    if isinstance(adata.X, csr_matrix):
+        number_per_cell = np.diff(adata.X.indptr)
+    elif isinstance(adata.X, csc_matrix):
+        cell_indices, counts = np.unique(adata.X.indices, return_counts=True)
+
+        # We take care of special case that an entire row has no values (implicit zero)
+        # and therefore omitted in the unique count of `adata.X.indices` by allocating
+        # a new array of size equal to the number of rows in the sparse matrix where each
+        # index in the row corresponds to a row index in the sparse matrix and then filling
+        # this new array with counts non-zero columns for a corresponding row index
+        #
+        # Example: 4X4 sparse matrix stored in column orientation where ROW 3 has no values.
+        # That is, ROW 3 is not stored in the `adata.X.indices` array:
+        # column_oriented_sparse_matrix = np.array([[1, 0, 0, 2], [0, 4, 1, 0], [0, 0, 0, 0], [0, 0, 5, 0]])
+        number_per_cell = np.zeros(adata.shape[0], dtype="int")
+        number_per_cell[cell_indices] = counts
+    elif isinstance(adata.X, np.ndarray):
+        number_per_cell = (adata.X > 0).sum(axis=1).A1
```

Comment:
```suggestion
        number_per_cell = (adata.X > 0).sum(axis=1).flatten()
```
Apparently `.A1` is not a valid option.

Reply:
will fix!